cpp/src/arrow/builder.cc

```diff
@@ -741,8 +741,26 @@ Status DictionaryBuilder<T, Scalar>::Append(const Scalar& value) {
   return Status::OK();
 }
 
-template <typename T, typename Scalar>
-Status DictionaryBuilder<T, Scalar>::DoubleTableSize() {
+template <typename T>
+Status DictionaryBuilder<T>::AppendArray(const Array& array) {
+  const NumericArray<T>& numeric_array = static_cast<const NumericArray<T>&>(array);
```

Comment:
You could use `PrimitiveArray` here and cast the `raw_data()` to `const Scalar*`